### PR TITLE
Oobranch e : add backend

### PR DESCRIPTION
**Pull request by antalk2**
Adds backend: manages ranges and data storage for citation groups.
      Provides cursor to write citation markers to.

### CODE REVIEW

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (koppor) [inline src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:42]: Please add a class comment, why this name was chosen 😇. We are currently preparing JabRef v5.4. See https://github.com/JabRef/jabref/milestone/23.

Maybe the overall changes are large enough to call for JabRef 6.0?

Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:42]: added
```
/**
 * Backend52, Codec52 and OODataModel.JabRef52 refer to the mode of storage, encoding and
 * what-is-stored in the document under JabRef version 5.2. These basically did not change up to
 * JabRef 5.4.
 */
```

> Maybe the overall changes are large enough to call for JabRef 6.0?

In the openoffice part most changes are not user-visible.

File: src/main/java/org/jabref/logic/openoffice/backend/NamedRangeReferenceMark.java
Comment (calixtus) [inline src/main/java/org/jabref/logic/openoffice/backend/NamedRangeReferenceMark.java:151]: Same here: Please avoid repeating the namespace / class in the method title.

**Review by calixtus — CHANGES_REQUESTED** [review]
Hi, thanks again for this incredible huge work of refactoring our open office / libre office interface.
I tried to make some comments on some obvious code style things and tried to some very few simplifications. Also IntelliJ provided two suggestions.
What do you think?

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:156]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:115]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:157]: renamed cgid to groupId
renamed cg to group

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:212]: I am not sure what to write here. The test and the message is not targeted to the end user,
but the programmer. Removed the message.

Comment (calixtus) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:212]: But the message is shown to the end user in case an exception occurs. Maybe something like "File is corrupt. Unable to parse." or something. Remember that exceptions may only be used for really exceptional conditions in the program, not as a substitute to control the program flow. https://github.com/HugoMatilla/Effective-JAVA-Summary#57-use-exceptions-only-for-exceptional-conditions

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:218]: changed to

```java
        /*
         * Backend52 uses reference marks to (1) mark the location of the citation in the text and (2) to encode
         * the citation keys and citation type in the name of the reference mark. The name of the reference mark
         * has to be unique in the document.
         */
        String markName = Codec52.getUniqueMarkName(new HashSet<>(citationStorageManager.nrmGetUsedNames(doc)),
                                                    citationKeys,
                                                    citationType);
```

Moved the `Collectors.joining(",")` part into `Codec52.getUniqueMarkName`

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:223]: changed to `numberOfCitations`

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:255]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:276]: > Is this something to be completed in one of the followup prs?

Not in the oobranch-[A-J] series.

Changed to 
```java
        case JabRef60:
            throw new IllegalStateException("createCitationGroup for JabRef60 is not implemented yet");
        default:
            throw new IllegalStateException("Unhandled dataModel in Backend52.createCitationGroup");
```
to align with other parts.


My idea here was to eventually separate Backend52 and Backend60 (or whatever it is going to be called) behind a hopefully common interface. In this line of thought I considered the `case JabRef60:` branches temporary
here, only existing to make it easier to compare with the Jabref52/Backend52 parts.

Backend60 (or handling Jabref60) is not finished (and not moved to a separate class).
Suggestions seemed to point towards minimally changing the encoding, and providing only an upgrade possibility from the old encoding (not editing or exporting to older representation).

This would allow us to drop Jabref52-specific handling from at least "Separate citations" and "Merge citations",
and move back code from Backend to the action implementation.

On the other hand there are several problems with the encoding used and I am afraid releasing a new backend
for each fix might result in too many variants to handle.

@koppor 
Lately [koppor suggested](https://github.com/JabRef/jabref/issues/2146#issuecomment-891385140)  to 
"use the same format as Zotero? So that OO documents can be used with Zotero and JabRef interchangeably"
and "My proposal would be: Switch to CSL only for JabRef 6.0"

This brings further changes affecting the representation in view. 
In short, I am not sure what is the next step after oobranch-J. Implement Jabref60/Backend60 with minimal changes,
or extend internal handling first to cover all parts of Zotero (prefix, suffix, ...) internally, then try to use zotero-like encoding in the document?

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:212]: I would write `assert pageInfos.size() == citationKeys.size()` if assertions were on.

The test is only needed because the function receives two lists and I could not express
statically that they should have the same number of elements. The concern is not
user input or files, just the caller providing data in the expected shape.
Might be an overkill, since the only caller `Frontend.createCitationGroup`,
as well as the only caller of `Frontend.createCitationGroup`: `UpdateCitationMarkers.createAndFillCitationGroup` already contain the same test.

I was thinking about changing the parameters of the function to contain 
`List<Pair<String,Optional<OOText>>> citationKeysAndPageInfos` or maybe
`List<Citation>` to clarify that here we expect an optional pageInfo for each key.

But I decided that the first may be too verbose (and vague), and the second may be confusing
because we would pass in Citation instances that have no representation in the document.
(Normally we read Citation and CitationGroup instances from the document,
or create them in the document here, along with creating the instance.)

The source of the data here is not the file. The only caller is `Frontend.createCitationGroup`,
which is called (only) from: `UpdateCitationMarkers.createAndFillCitationGroup`.

This latter is called from 3 locations:

- (1) `EditInsert.insertCitationGroup` which still receives a single pageInfo from the GUI which it embeds in a list
using `OODataModel.fakePageInfos(pageInfo, nEntries);`
-  (2) `EditMerge.mergeCitationGroups` which calls the backend (through Frontend): 
  `List<Optional<OOText>> pageInfos = fr.backend.combinePageInfos(cgs);` 
-  (3) `EditSeparate.separateCitations` which removes a citation group and creates a citation group for each citation

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:288]: Not off, just zero. Changed to 4

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:290]: renamed cgPageInfos to pageInfos
renamed cgPageInfo to singlePageInfo

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:295]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:85]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:217]: ok. (in Codec52.getUniqueMarkName now)

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (calixtus) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:276]: ah ok, I think this should be discussed explicitly. But discussing the future development of this should not stop us from integrating this oo pr series.

File: src/main/java/org/jabref/logic/openoffice/backend/Backend52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Backend52.java:329]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/Codec52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Codec52.java:60]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/Codec52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Codec52.java:73]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/NamedRangeManagerReferenceMark.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/NamedRangeManagerReferenceMark.java:19]: ### interface NamedRangeManager
nrmCreate -> createNamedRange
nrmGetUsedNames -> getUsedNames
nrmGetFromDocument -> getNamedRangeFromDocument

### interface NamedRange

added comment to clarify intent
```java
/**
 * NamedRange (with NamedRangeManager) attempts to provide a common interface for working with
 * reference mark based and bookmark based text ranges to be used as locations to fill with citation
 * markers. LibreOffice supports name-based lookup and listing names for both (hence the name).
 *
 * Note: currently only implemented for refence marks (in NamedRangeReferenceMark and
 *       NamedRangeManagerReferenceMark). 
 * 
 */
```

nrGetRangeName -> getRangeName
nrGetMarkRange -> getMarkRange
nrGetRawCursor -> getRawCursor
nrGetFillCursor -> getFillCursor
nrCleanFillCursor -> cleanFillCursor
nrRemoveFromDocument -> removeFromDocument

File: src/main/java/org/jabref/logic/openoffice/backend/Codec52.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/Codec52.java:140]: ok

File: src/main/java/org/jabref/model/openoffice/backend/NamedRange.java
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/backend/NamedRange.java:16]: Done above. The intent was to make it easy to search references.

File: src/main/java/org/jabref/model/openoffice/backend/NamedRangeManager.java
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/backend/NamedRangeManager.java:15]: Opted for createNamedRange above. (I see Intellij can find call references after it finished indexing)

File: src/main/java/org/jabref/logic/openoffice/backend/NamedRangeReferenceMark.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/NamedRangeReferenceMark.java:207]: ok

File: src/main/java/org/jabref/logic/openoffice/backend/NamedRangeReferenceMark.java
Comment (antalk2) [reply src/main/java/org/jabref/logic/openoffice/backend/NamedRangeReferenceMark.java:151]: ok.

File: src/main/java/org/jabref/model/openoffice/backend/NamedRange.java
Comment (calixtus) [reply src/main/java/org/jabref/model/openoffice/backend/NamedRange.java:16]: Yes I see that and appreciate the thinking, but we are not in the 90s writing our code with vi. IntelliJ for Instance provides the class names on the right hand side of the search window and a brief context at the bottom. I'm sure netbeans, vs/vscode and eclipse provide something similar. 😉

**Comment by calixtus** [comment]
Looks good to me! Thanks. After a second review we can merge.

**Review by calixtus — APPROVED** [review]

File: src/main/java/org/jabref/logic/openoffice/backend/Codec52.java
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/openoffice/backend/Codec52.java:113]: i is normally used only in loops, but for the moment it doesn't matter. You can fix this in a later PR.

**Review by Siedlerchr — APPROVED** [review]
Just some minor nitpicking things, but nothing important. So I merge ths now.
A general hint: Please use more descriptive variable names where possibe

**Pull request merged**
